service/history/workflow/mutable_state_impl.go

```diff
@@ -2567,11 +2572,18 @@ func (ms *MutableStateImpl) addCompletionCallbacks(
 			}
 		}
 		machine := callbacks.NewCallback(event.EventTime, callbacks.NewWorkflowClosedTrigger(), persistenceCB)
-		// Use the start event version and ID as part of the callback ID to ensure that callbacks have unique
-		// IDs that are deterministically created across clusters.
-		// TODO: Replicate the state machine state and allocate a uuid there instead of relying on history event
-		// replication.
-		id := fmt.Sprintf("%d-%d-%d", event.GetVersion(), event.GetEventId(), idx)
+		id := ""
+		// This is for backwards compatibility: callbacks was initially only attached when the workflow
+		// execution started, but now it can be attached while it's running.
+		if event.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED {
+			// Use the start event version and ID as part of the callback ID to ensure that callbacks have unique
+			// IDs that are deterministically created across clusters.
+			// TODO: Replicate the state machine state and allocate a uuid there instead of relying on history event
+			// replication.
```

Comment:
You can remove the TODO here, this will be made obsolete with CHASM and state based replication.